app/controllers/checkouts_controller.rb

```diff
@@ -18,9 +26,49 @@ def create
     end
   end
 
+  def return
+    @checkout = Checkout.find(params[:id])
+  end
+
+  def complete_return
+    @checkout = Checkout.find(params[:id])
+    @checkout.returned_at = params[:checkout][:returned_at]
+
+    if @checkout.save
+      redirect_to checkouts_path, notice: 'Book returned successfully.'
+    else
+      render :return
+    end
+  end
+  
+  def edit
+    @checkout = Checkout.find(params[:id])
+  end
+
+  def update
+    @checkout = Checkout.find(params[:id])
+
+    if @checkout.update(checkout_params)
+      redirect_to checkouts_path, notice: 'Checkout updated successfully.'
+    else
+      render :edit
+    end
+  end
+
+  def destroy
+    @checkout = Checkout.find(params[:id])
+    @checkout.destroy
+
+    redirect_to checkouts_path, notice: 'Checkout deleted successfully.'
+  end
+
   private
 
   def checkout_params
     params.require(:checkout).permit(:book_id, :patron_id, :checked_out_at, :due_date, :returned_at)
```

Comment:
checkout_at and returned_at should be returned from front-end, instead set the current timestamp in controller for these values